### PR TITLE
[Common] update caps dimension properties

### PR DESCRIPTION
**Pull request by niley7464**
This patch adds update caps dimension properties.
`update_caps_dimension` uses peer_caps for negotiation and treats `a:b:c` and `a:b:c:1` eqaul.

See also: #3937 

Signed-off-by: Yelin Jeong <yelini.jeong@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c [hunk 1]
```diff
@@ -956,6 +1019,9 @@ gst_tensor_pad_caps_from_config (GstPad * pad, const GstTensorsConfig * config)
   /* other/tensor */
   if (config->info.num_tensors == 1 && _peer_has_tensor_caps (pad)) {
     caps = _get_tensor_caps (config);
+    if ((peer_caps = gst_pad_peer_query_caps (pad, NULL))) {
+      update_caps_dimension (config, caps, peer_caps);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/nnstreamer_plugin_api_impl.c:1023]: line 1023 and 1034
add unref after usage: `gst_caps_unref (peer_caps);`

Comment (niley7464) [reply gst/nnstreamer/nnstreamer_plugin_api_impl.c]: updated! Thank you :)

File: tests/nnstreamer_converter/runTest.sh [hunk 2]
```diff
@@ -83,6 +83,9 @@ gstTest "--gst-plugin-path=${PATH_TO_PLUGIN} audiotestsrc num-buffers=1 samplesp
 gstTest "--gst-plugin-path=${PATH_TO_PLUGIN} multifilesrc location=\"testsequence_%1d.png\" index=0 caps=\"image/png,framerate=\(fraction\)30/1\" ! pngdec ! videoconvert ! tensor_converter ! filesink location=\"testcase08.log\"" 6-1 0 0 $PERFORMANCE
 callCompareTest testcase08.golden testcase08.log 6-2 "PNG Stream Test" 0 0
 
+# Dimension declaration test case
+gstTest "--gst-plugin-path=${PATH_TO_PLUGIN} videotestsrc num-buffers=1 ! videoscale ! videoconvert ! video/x-raw,format=RGB,framerate=30/1,height=300,width=300 ! tensor_converter ! other/tensor,dimension=3:300:300,types=uint8,framerate=30/1,format=static ! fakesink" 7 0 0 $PERFORMANCE
```
Comment (anyj0527) [inline tests/nnstreamer_converter/runTest.sh:87]: typo "types"->"type".
How about add a simple case for `other/tensors`?

File: gst/nnstreamer/include/nnstreamer_plugin_api.h [hunk 3]
```diff
@@ -100,5 +100,19 @@ gst_tensor_meta_info_parse_memory (GstTensorMetaInfo * meta, GstMemory * mem);
...
+ * @param peer_caps caps to compare
+ */
+extern void update_caps_dimension (GstCaps * caps, GstCaps * peer_caps);
+
+/**
+ * @brief Update caps dimensions for negotiation
+ * @param caps caps to compare and update
+ * @param peer_caps caps to compare
+ */
+extern void update_caps_dimensions (GstCaps * caps, GstCaps * peer_caps);
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api.h:115]: Please keep the code style consistent.

File: tests/nnstreamer_converter/runTest.sh (hunk 2 above)
Comment (niley7464) [reply tests/nnstreamer_converter/runTest.sh:87]: I will make more test cases using tensor_* elements 👍

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c [hunk 4]
```diff
@@ -923,6 +923,83 @@ _peer_is_flexible_tensor_caps (GstPad * pad)
...
+void
+update_caps_dimension (GstCaps * caps, GstCaps * peer_caps)
+{
+  GstStructure *structure;
+  GstStructure *structure_peer;
+  tensor_dim dim, dim_peer;
+  const gchar *dim_str;
+  const gchar *dim_str_peer;
+  guint i;
+
```
Comment (songgot) [inline gst/nnstreamer/nnstreamer_plugin_api_impl.c:940]: How about checking for NULL?

g_return_val_if_fail (caps != NULL, FALSE);
g_return_val_if_fail (peer_caps != NULL, FALSE);

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c [hunk 5]
```diff
@@ -923,6 +923,83 @@ _peer_is_flexible_tensor_caps (GstPad * pad)
...
+  if (gst_caps_get_size (caps) == 0 || gst_caps_get_size (peer_caps) == 0)
+    return;
+
+  structure = gst_caps_get_structure (caps, 0);
+  dim_str = gst_structure_get_string (structure, "dimension");
+  gst_tensor_parse_dimension (dim_str, dim);
```
Comment (songgot) [inline gst/nnstreamer/nnstreamer_plugin_api_impl.c:946]: dim_str can be null in some cases if there is no dimension in structure.
how about check retrun value ?
if (!gst_tensor_parse_dimension (dim_str, dim)) {
  /* error log */
  return;
}

File: gst/nnstreamer/include/nnstreamer_plugin_api.h (hunk 3 above)
Comment (niley7464) [reply gst/nnstreamer/include/nnstreamer_plugin_api.h]: updated.
Thank you!

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c (hunk 4 above)
Comment (niley7464) [reply gst/nnstreamer/nnstreamer_plugin_api_impl.c:939]: updated to check NULL value
Thank you!

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c (hunk 5 above)
Comment (niley7464) [reply gst/nnstreamer/nnstreamer_plugin_api_impl.c]: Updated to parse dimension when structure has dimension properties only.
Thank you!

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c [hunk 6]
```diff
@@ -923,6 +923,188 @@ _peer_is_flexible_tensor_caps (GstPad * pad)
...
+  g_return_val_if_fail (gst_structure_has_field (structure2, "dimension"),
+      FALSE);
+
+  dim_str1 = gst_structure_get_string (structure1, "dimension");
+  gst_tensor_parse_dimension (dim_str1, dim1);
+
+  dim_str2 = gst_structure_get_string (structure2, "dimension");
+  gst_tensor_parse_dimension (dim_str2, dim2);
+
+  if (!_is_tensor_dim_same (dim1, dim2))
```
Comment (jaeyun-jung) [inline gst/nnstreamer/nnstreamer_plugin_api_impl.c:962]: minor: simply `return _is_tensor_dim_same (dim1, dim2);`

File: gst/nnstreamer/include/nnstreamer_plugin_api.h [hunk 7]
```diff
@@ -100,5 +100,22 @@ gst_tensor_meta_info_parse_memory (GstTensorMetaInfo * meta, GstMemory * mem);
 extern GstMemory *
 gst_tensor_meta_info_append_header (GstTensorMetaInfo * meta, GstMemory * mem);
 
+/**
+ * @brief Update caps dimension for negotiation
+ * @param caps caps to compare and update
+ * @param peer_caps caps to compare
+ */
+extern void
+update_caps_dimension (GstCaps *caps, GstCaps *peer_caps);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/include/nnstreamer_plugin_api.h:109]: minor: name rule - prefix
```
gst_tensor_caps_update_dimension (GstCaps *caps, GstCaps *peer);
gst_tensor_caps_can_intersect (GstCaps *caps1, GstCaps *caps2);
```

File: gst/nnstreamer/nnstreamer_plugin_api_impl.c (hunk 6 above)
Comment (niley7464) [reply gst/nnstreamer/nnstreamer_plugin_api_impl.c]: Updated.
Thank you :)

File: gst/nnstreamer/include/nnstreamer_plugin_api.h (hunk 7 above)
Comment (niley7464) [reply gst/nnstreamer/include/nnstreamer_plugin_api.h]: Updated.
Thank you :)

File: gst/nnstreamer/elements/gsttensor_transform.c [hunk 8]
```diff
@@ -1989,12 +1989,27 @@ gst_tensor_transform_transform_caps (GstBaseTransform * trans,
...
     gst_caps_unref (result);
     result = intersection;
+
+    if (direction == GST_PAD_SINK)
+      pad = GST_BASE_TRANSFORM_SRC_PAD (filter);
+    else
+      pad = GST_BASE_TRANSFORM_SINK_PAD (filter);
+
+    if ((peer_caps = gst_pad_peer_query_caps (pad, NULL))) {
+      structure = gst_caps_get_structure (peer_caps, 0);
```
Comment (songgot) [inline gst/nnstreamer/elements/gsttensor_transform.c:2009]: structure is not used. so can we remove this ?

Comment (niley7464) [reply gst/nnstreamer/elements/gsttensor_transform.c]: Removed not used structure
Thank you :)

**Review by songgot — APPROVED** [review]
LGTM

**Review by myungjoo — APPROVED** [review]

**Pull request merged**